download/download.go

```diff
@@ -0,0 +1,181 @@
+		return "", renameErr
+	}
+
+	log.Info(context.Background(), "mongod downloaded and stored in cache", log.Data{"filename": cfg.cachePath, "ellapsed": time.Since(downloadStartTime).String()})
+
+	return cfg.cachePath, nil
+}
+
+// downloadFile downloads the file from the given url and stores it in a temporary file.
+// It returns the path to the temporary file where it has been downloaded
+func downloadFile(urlStr string) (afero.File, error) {
+	log.Info(context.Background(), "Downloading file", log.Data{"url": urlStr})
+
+	resp, httpGetErr := http.Get(urlStr)
+	if httpGetErr != nil {
+		return nil, httpGetErr
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("invalid status code %d", resp.StatusCode)
+	}
+
+	defer resp.Body.Close()
+
+	tgzTempFile, tmpFileErr := afs.TempFile("", "")
+	if tmpFileErr != nil {
+		return nil, tmpFileErr
+	}
+
+	_, copyErr := io.Copy(tgzTempFile, resp.Body)
+	if copyErr != nil {
+		_ = tgzTempFile.Close()
+		_ = afs.Remove(tgzTempFile.Name())
+		return nil, copyErr
+	}
+
+	log.Info(context.Background(), "Downloaded to temp file", log.Data{"file": tgzTempFile.Name(), "url": urlStr})
+	return tgzTempFile, nil
+}
+
+// extractMongoBin extracts the mongod executable file
+// from the given tarball to a temporary file.
+// It returns the path to the extracted file
+func extractMongoBin(tgzTempFile afero.File) (string, error) {
+	_, seekErr := tgzTempFile.Seek(0, 0)
+	if seekErr != nil {
+		log.Error(context.Background(), "error seeking back to start of file", seekErr)
+		return "", seekErr
+	}
+
+	gzReader, gzErr := gzip.NewReader(tgzTempFile)
+	if gzErr != nil {
+		log.Error(context.Background(), "error intializing gzip reader", gzErr, log.Data{"file": tgzTempFile.Name()})
+		return "", gzErr
+	}
+
+	tarReader := tar.NewReader(gzReader)
+
+	for {
+		nextFile, tarErr := tarReader.Next()
+		if tarErr == io.EOF {
+			return "", fmt.Errorf("did not find a mongod binary in the tar file")
+		}
+		if tarErr != nil {
+			log.Error(context.Background(), "error reading from tar file", tarErr, log.Data{"file": tgzTempFile.Name()})
+			return "", tarErr
+		}
+
+		if strings.HasSuffix(nextFile.Name, "bin/mongod") {
+			break
+		}
+	}
+
+	// Extract to a temp file first, then copy to the destination, so we get
+	// atomic behavior if there's multiple parallel downloaders
+	mongodTmpFile, tmpFileErr := afs.TempFile("", "")
+	if tmpFileErr != nil {
+		log.Error(context.Background(), "error creating temp file for mongod", tmpFileErr)
+		return "", tmpFileErr
+	}
+	defer func() {
+		_ = mongodTmpFile.Close()
+	}()
+
+	_, writeErr := io.Copy(mongodTmpFile, tarReader)
+	if writeErr != nil {
+		log.Error(context.Background(), "error writing mongod binary", writeErr, log.Data{"filename": mongodTmpFile.Name()})
+		return "", writeErr
+	}
+
+	_ = mongodTmpFile.Close()
+
+	chmodErr := afs.Chmod(mongodTmpFile.Name(), 0755)
```

Comment:
consider moving the chmod to after the verify

Reply:
I believe this is already after the verify? The `extractMongoBin` (this function) [is called right after passing verification](https://github.com/ONSdigital/dp-mongodb-in-memory/blob/ade2bebd4549a33ed374931f0521b6bdfa21ed73/download/download.go#L59-L68)